# Task 1 Importing and Cleaning Data

First we need to import the required libraries for the  project.

In [1]:
#importing required libraries
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup
import requests

### Import Data
1. I retrieved the data from the source website using the requests library.
2. Use Beautiful Soup to find instances of 'table' on the website.
3. Use pandas to convert it into a dataframe.

In [2]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M' #identify url
source = requests.get(url).text #scrape the website
soup = BeautifulSoup(source, 'lxml') #parse the code
table = soup.find_all('table') #find instances of table
df = pd.read_html(str(table))[0] #convert to data frame

### Identify  Missing Values
Before cleaning the data had 180 rows.
After using `.describe()` I could see that there were 180 unique values for Postal Code but  77 "Not Assigned" Boroughs.

In [3]:
df.describe()

Postal code       Borough Neighborhood
count          180           180          103
unique         180            11           98
top            M6T  Not assigned    Downsview
freq             1            77            4

### Deal with missing values
After removing those I could see that I now had 103 rows.

In [4]:
print("Before Cleaning",df.shape)
df = df[df['Borough']!="Not assigned"] #Remove Not Assigned from Borough
print("Removed Not Assigned",df.shape)

Before Cleaning (180, 3)
Removed Not Assigned (103, 3)


The assignment instructions stated:
>If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

However, when I tried to find any unassigned there weren't any!

In [5]:
df[['Neighborhood']].isna().groupby('Neighborhood').count()

Empty DataFrame
Columns: []
Index: [False]

In [6]:
#Commented out as the code is redundant. There is no data that meets those requirements for cleaning
    #df = df[df['Neighborhood']!=np.nan]
    #print("Removed Nulls",df.shape)

### Correct the data format
Next, I reformatted the Neighborhood column so that the " /" were replaced with ",".\
I then capitalised the "c" in Postal code.
I then re-indexed to reflect the missing columns.

In [7]:
df['Neighborhood']  = df['Neighborhood'].str.replace(" /", ",")  # replaced / with ,
df.columns=["Postal Code", "Borough", "Neighborhood"]
df.reset_index(drop=True,inplace=True)
df.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

### Commentary
Whilst there are expected duplicates in the borough column, I can see there are no duplicates in the postal code column.\
There are however, duplicates in the Neighborhood column.

In [8]:
df.describe(include='all')

Postal Code     Borough Neighborhood
count          103         103          103
unique         103          10           98
top            M1T  North York    Downsview
freq             1          24            4

### Assumptions
I will assume that those duplicates are due to a Neighborhood spanning multiple postcodes but only **a single borough per Neighborhood**.\
The code below also shows that there are no hidden duplicates of Borough, through different spellings or formats.

In [9]:
df['Borough'].value_counts()

North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East York            5
York                 5
East Toronto         5
Mississauga          1
Name: Borough, dtype: int64

#### Let's see  what the final data frame looks like.

In [10]:
df.shape

(103, 3)

# Task 2 - Importing Geospatial Data

Data imported from csv directly into data frame.

In [11]:
geo = pd.read_csv("https://cocl.us/Geospatial_data")
geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Next, we need to do a left join of df and geo dataframes.

In [12]:
df = df.merge(geo,on="Postal Code")
df.head(11)

Postal Code           Borough                                 Neighborhood  \
0          M3A        North York                                    Parkwoods   
1          M4A        North York                             Victoria Village   
2          M5A  Downtown Toronto                    Regent Park, Harbourfront   
3          M6A        North York             Lawrence Manor, Lawrence Heights   
4          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5          M9A         Etobicoke                             Islington Avenue   
6          M1B       Scarborough                               Malvern, Rouge   
7          M3B        North York                                    Don Mills   
8          M4B         East York              Parkview Hill, Woodbine Gardens   
9          M5B  Downtown Toronto                     Garden District, Ryerson   
10         M6B        North York                                    Glencairn   

     Latitude  Longitude  
0   43.753259 -79.329656  
1   43.725882 -79.315572  
2   43.654260 -79.360636  
3   43.718518 -79.464763  
4   43.662301 -79.389494  
5   43.667856 -79.532242  
6   43.806686 -79.194353  
7   43.745906 -79.352188  
8   43.706397 -79.309937  
9   43.657162 -79.378937  
10  43.709577 -79.445073

# Task 3 - Segment and Cluster Neighborhoods

### Map the Neighborhood

First let's import more libraries.

In [13]:
import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json #
from pandas.io.json import json_normalize
import matplotlib.cm as cm
import matplotlib.colors as colors
# import k-means from clustering stage
from sklearn.cluster import KMeans
# map rendering library
!pip install folium
import folium 
print('Libraries imported.')

Libraries imported.


Next,take slice the data frame to only include boroughs that contain the word Toronto.

In [14]:
df_dt = df[df['Borough'] =="Downtown Toronto"]
df_ct = df[df['Borough'] =="Central Toronto"]
df_wt = df[df['Borough'] =="West Toronto"]
df_et = df[df['Borough'] =="East Toronto"]
tor = df_dt.append([df_ct,df_wt,df_et]).reset_index(drop=True)

In [15]:
# create map of Toronto using latitude and longitude values
latitude = 43.6632
longitude = -79.3832
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(tor['Latitude'],tor['Longitude'], tor['Borough'], tor['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

# Gather Data from Foursquare and Transform

Create hidden cell containing Foursquare Credentials

In [16]:
# The code was removed by Watson Studio for sharing.

This text is the result of a Hidden Cell with Foursquare Credentials


Define a function to retrieve the nearby venues, given a longitude and latitude.

In [17]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Create a data frame that contains the venue information for each neighbourhood.

In [18]:
tor_venues = getNearbyVenues(names=tor['Neighborhood'],
                                   latitudes=tor['Latitude'],
                                   longitudes=tor['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
North Toronto West
The Annex, North Midtown, Yorkville
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High Park, The Junction South
Parkdale, Roncesvalles
Runnymede, Swansea
The Beaches
The

Check the data frame looks as expected.

In [19]:
print(tor_venues.shape)
tor_venues.head()

(1612, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4      Morning Glory Cafe       43.653947       -79.361149   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4       Breakfast Spot

**Explore** how many venues in each neighborhood.

In [20]:
tor_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            58   
Brockton, Parkdale Village, Exhibition Place                           22   
Business reply mail Processing CentrE                                  17   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     67   
Christie                                                               17   
Church and Wellesley                                                   71   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             36   
Davisville North                                                        7   
Dufferin, Dovercourt Village                                           16   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                5   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
India Bazaar, The Beaches West                                         19   
Kensington Market, Chinatown, Grange Park                              56   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               41   
Moore Park, Summerhill East                                             1   
North Toronto West                                                     19   
Parkdale, Roncesvalles                                                 13   
Queen's Park, Ontario Provincial Government                            38   
Regent Park, Harbourfront                                              48   
Richmond, Adelaide, King                                               93   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     37   
St. James Town                                                         84   
St. James Town, Cabbagetown                                            44   
Stn A PO Boxes                                                         96   
Studio District                                                        40   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    22   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             58   
Brockton, Parkdale Village, Exhibition Place                            22   
Business reply mail Processing CentrE                                   17   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      67   
Christie                                                                17   
Church and Wellesley                                                    71   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              36   
Davisville North                    

In [21]:
print('There are {} uniques categories.'.format(len(tor_venues['Venue Category'].unique())))

There are 230 uniques categories.


One Hot Encoding of venues to allow for clustering of categorical data.

In [22]:
# one hot encoding
tor_onehot = pd.get_dummies(tor_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
tor_onehot['Neighborhood'] = tor_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [tor_onehot.columns[-1]] + list(tor_onehot.columns[:-1])
tor_onehot = tor_onehot[fixed_columns]

tor_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Lounge  Airport Service  \
0            0        0                   0               0                0   
1            0        0                   0               0                0   
2            0        0                   0               0                0   
3            0        0                   0               0                0   
4            0        0                   0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Auto Workshop  \
0           0                    0                 0              0   
1           0                    0                 0              0   
2           0                    0                 0              0   
3           0                    0                 0              0   
4           0                    0                 0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       1     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bike Rental / Bike Share  Bistro  Boat or Ferry  \
0                   0                         0       0              0   
1                   0                         0       0              0   
2                   0                         0       0              0   
3                   0                         0       0              0   
4                   0                         0       0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               1        0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  Butcher  \
0                0         0             0              0         0        0   
1                0         0             0              0         0        0   
2                0         0             0              0         0        0   
3                0         0             0              0         0        0   
4                0         0             0              0         0        0   

   Café  Cajun / Creole Restaurant  Candy Store  Caribbean Restaurant  \
0     0                          0            0                     0   
1     0                          0            0                     0   
2     0                          0            0                     0   
3     0                          0            0              

In [23]:
tor_onehot.shape

(1612, 230)

Group venues by neighborhood and get a mean frequency for each venue category.

In [24]:
tor_grouped = tor_onehot.groupby('Neighborhood').mean().reset_index()
tor_grouped

Neighborhood  Yoga Studio  Airport  \
0                                         Berczy Park     0.000000   0.0000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000   0.0000   
2               Business reply mail Processing CentrE     0.058824   0.0000   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   0.0625   
4                                  Central Bay Street     0.014925   0.0000   
5                                            Christie     0.000000   0.0000   
6                                Church and Wellesley     0.028169   0.0000   
7                      Commerce Court, Victoria Hotel     0.000000   0.0000   
8                                          Davisville     0.000000   0.0000   
9                                    Davisville North     0.000000   0.0000   
10                       Dufferin, Dovercourt Village     0.000000   0.0000   
11             First Canadian Place, Underground city     0.000000   0.0000   
12                           Forest Hill North & West     0.000000   0.0000   
13                           Garden District, Ryerson     0.000000   0.0000   
14  Harbourfront East, Union Station, Toronto Islands     0.000000   0.0000   
15                      High Park, The Junction South     0.000000   0.0000   
16                     India Bazaar, The Beaches West     0.000000   0.0000   
17          Kensington Market, Chinatown, Grange Park     0.000000   0.0000   
18                                      Lawrence Park     0.000000   0.0000   
19                           Little Portugal, Trinity     0.024390   0.0000   
20                        Moore Park, Summerhill East     0.000000   0.0000   
21                                 North Toronto West     0.052632   0.0000   
22                             Parkdale, Roncesvalles     0.000000   0.0000   
23        Queen's Park, Ontario Provincial Government     0.026316   0.0000   
24                          Regent Park, Harbourfront     0.020833   0.0000   
25                           Richmond, Adelaide, King     0.000000   0.0000   
26                                           Rosedale     0.000000   0.0000   
27                                           Roselawn     0.000000   0.0000   
28                                 Runnymede, Swansea     0.027027   0.0000   
29                                     St. James Town     0.000000   0.0000   
30                        St. James Town, Cabbagetown     0.000000   0.0000   
31                                     Stn A PO Boxes     0.010417   0.0000   
32                                    Studio District     0.025000   0.0000   
33  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   0.0000   
34                The Annex, North Midtown, Yorkville     0.000000   0.0000   
35                                        The Beaches     0.000000   0.0000   
36                       The Danforth West, Riverdale     0.023256   0.0000   
37           Toronto Dominion Centre, Design Exchange     0.000000   0.0000   
38                     University of Toronto, Harbord     0.029412   0.0000   

    Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0               0.0000           0.000           0.0000             0.000   
1               0.0000           0.000           0.0000             0.000   
2               0.0000           0.000           0.0000             0.000   
3               0.0625           0.125           0.1875             0.125   
4               0.0000           0.000           0.0000             0.000   
5               0.0000           0.000           0.0000             0.000   
6               0.0000           0.000           0.0000             0.000   
7               0.0000           0.000           0.0000             0.000   
8               0.0000           0.000           0.0000             0.000   
9               0.0000           0.000           0.0000             0.000   
10              0.0000           

In [25]:
tor_grouped.shape

(39, 230)

**Explore data** to see if there is a significant difference in the first 5 venue categories.

In [26]:
num_top_venues = 5

for hood in tor_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = tor_grouped[tor_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2  Italian Restaurant  0.03
3              Bakery  0.03
4                Café  0.03


----Brockton, Parkdale Village, Exhibition Place----
                venue  freq
0                Café  0.14
1         Coffee Shop  0.09
2      Breakfast Spot  0.09
3                 Gym  0.05
4  Italian Restaurant  0.05


----Business reply mail Processing CentrE----
         venue  freq
0  Yoga Studio  0.06
1   Restaurant  0.06
2      Brewery  0.06
3   Smoke Shop  0.06
4         Park  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                venue  freq
0     Airport Service  0.19
1      Airport Lounge  0.12
2    Airport Terminal  0.12
3         Coffee Shop  0.06
4  Airport Food Court  0.06


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.19
1   Italian Restaurant  0.06
2             

Define a function that gathers the most common venues for a given neighborhood.

In [27]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Collect the 15 most common venue categories in each neighbourhood.

In [28]:
num_top_venues = 15

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = tor_grouped['Neighborhood']

for ind in np.arange(tor_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(tor_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2              Business reply mail Processing CentrE           Yoga Studio   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar           Cheese Shop        Farmers Market   
1           Coffee Shop        Breakfast Spot                   Gym   
2                   Spa         Garden Center                Garden   
3        Airport Lounge      Airport Terminal      Sculpture Garden   
4    Italian Restaurant        Sandwich Place                  Café   

  5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0    Seafood Restaurant                Bakery             Restaurant   
1                Bakery     Convenience Store  Performing Arts Venue   
2  Fast Food Restaurant        Farmers Market     Light Rail Station   
3                 Plane               Airport     Airport Food Court   
4           Salad Place   Fried Chicken Joint        Bubble Tea Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
0    Italian Restaurant                  Café               Beer Bar   
1             Pet Store          Climbing Gym             Restaurant   
2            Comic Shop           Pizza Place                Butcher   
3   Rental Car Location              Boutique        Harbor / Marina   
4          Burger Joint   Japanese Restaurant         Ice Cream Shop   

  11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
0                   Park                  Diner           Concert Hall   
1          Burrito Place     Italian Restaurant           Intersection   
2             Restaurant          Burrito Place                Brewery   
3          Boat or Ferry            Coffee Shop       Doner Restaurant   
4        Thai Restaurant       Department Store          Metro Station   

  14th Most Common Venue      15th Most Common Venue  
0          Shopping Mall                      Bistro  
1                Stadium                         Bar  
2             Skate Park                  Smoke Shop  
3             Donut Shop         Dumpling Restaurant  
4      Korean Restaurant  Modern European Restaurant

# Cluster Neigborhoods

In [29]:
# set number of clusters
kclusters = 5

tor_grouped_clustering = tor_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(tor_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 4, 0, 2, 0,
       0, 0, 0, 0, 3, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [30]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
tor_merged = tor

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
tor_merged = tor_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
tor_merged # check the last columns!

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M5E  Downtown Toronto   
5          M5G  Downtown Toronto   
6          M6G  Downtown Toronto   
7          M5H  Downtown Toronto   
8          M5J  Downtown Toronto   
9          M5K  Downtown Toronto   
10         M5L  Downtown Toronto   
11         M5S  Downtown Toronto   
12         M5T  Downtown Toronto   
13         M5V  Downtown Toronto   
14         M4W  Downtown Toronto   
15         M5W  Downtown Toronto   
16         M4X  Downtown Toronto   
17         M5X  Downtown Toronto   
18         M4Y  Downtown Toronto   
19         M4N   Central Toronto   
20         M5N   Central Toronto   
21         M4P   Central Toronto   
22         M5P   Central Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M4S   Central Toronto   
26         M4T   Central Toronto   
27         M4V   Central Toronto   
28         M6H      West Toronto   
29         M6J      West Toronto   
30         M6K      West Toronto   
31         M6P      West Toronto   
32         M6R      West Toronto   
33         M6S      West Toronto   
34         M4E      East Toronto   
35         M4K      East Toronto   
36         M4L      East Toronto   
37         M4M      East Toronto   
38         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                           Regent Park, Harbourfront  43.654260 -79.360636   
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
2                            Garden District, Ryerson  43.657162 -79.378937   
3                                      St. James Town  43.651494 -79.375418   
4                                         Berczy Park  43.644771 -79.373306   
5                                  Central Bay Street  43.657952 -79.387383   
6                                            Christie  43.669542 -79.422564   
7                            Richmond, Adelaide, King  43.650571 -79.384568   
8   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
9            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
10                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
11                     University of Toronto, Harbord  43.662696 -79.400049   
12          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
13  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420   
14                                           Rosedale  43.679563 -79.377529   
15                                     Stn A PO Boxes  43.646435 -79.374846   
16                        St. James Town, Cabbagetown  43.667967 -79.367675   
17             First Canadian Place, Underground city  43.648429 -79.382280   
18                               Church and Wellesley  43.665860 -79.383160   
19                                      Lawrence Park  43.728020 -79.388790   
20                                           Roselawn  43.711695 -79.416936   
21                                   Davisville North  43.712751 -79.390197   
22                           Forest Hill North & West  43.696948 -79.411307   
23                                 North Toronto West  43.715383 -79.405678   
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678   
25                                         Davisville  43.704324 -79.388790   
26                        Moore Park, Summerhill East  43.689574 -79.383160   
27  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049   
28                       Dufferin, Dovercourt Village  43.669005 -79.442259   
29                           Little Portugal, Trinity  43.647927 -79.419750   
30       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
31                      High Park,

In [31]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.viridis(np.linspace(0, 1, len(ys)))
csch = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tor_merged['Latitude'], tor_merged['Longitude'], tor_merged['Neighborhood'], tor_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=csch[cluster-1],
        fill=True,
        fill_color=csch[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)

map_clusters

## Cluster 1 - Coffee Shops and Cafes

In [32]:
tor_merged.loc[tor_merged['Cluster Labels'] == 0, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0        Clothing Store   
3   Downtown Toronto               0           Coffee Shop   
4   Downtown Toronto               0           Coffee Shop   
5   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0         Grocery Store   
7   Downtown Toronto               0           Coffee Shop   
8   Downtown Toronto               0           Coffee Shop   
9   Downtown Toronto               0           Coffee Shop   
10  Downtown Toronto               0           Coffee Shop   
11  Downtown Toronto               0                  Café   
12  Downtown Toronto               0                  Café   
13  Downtown Toronto               0       Airport Service   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   
17  Downtown Toronto               0           Coffee Shop   
18  Downtown Toronto               0      Sushi Restaurant   
21   Central Toronto               0                   Gym   
23   Central Toronto               0        Clothing Store   
24   Central Toronto               0        Sandwich Place   
25   Central Toronto               0        Sandwich Place   
27   Central Toronto               0                   Pub   
28      West Toronto               0                Bakery   
29      West Toronto               0                   Bar   
30      West Toronto               0                  Café   
31      West Toronto               0         Grocery Store   
32      West Toronto               0             Gift Shop   
33      West Toronto               0           Pizza Place   
34      East Toronto               0     Health Food Store   
35      East Toronto               0      Greek Restaurant   
36      East Toronto               0  Fast Food Restaurant   
37      East Toronto               0                  Café   
38      East Toronto               0           Yoga Studio   

   2nd Most Common Venue        3rd Most Common Venue  \
0                    Pub                       Bakery   
1                  Diner             Sushi Restaurant   
2            Coffee Shop                         Café   
3                   Café                 Cocktail Bar   
4           Cocktail Bar                  Cheese Shop   
5     Italian Restaurant               Sandwich Place   
6                   Café                         Park   
7                   Café                   Restaurant   
8               Aquarium           Italian Restaurant   
9                  Hotel                         Café   
10            Restaurant                         Café   
11                Bakery                          Bar   
12           Coffee Shop        Vietnamese Restaurant   
13        Airport Lounge             Airport Terminal   
15    Italian Restaurant                         Café   
16                Market           Italian Restaurant   
17                  Café                   Restaurant   
18           Coffee Shop          Japanese Restaurant   
21                 Hotel                         Park   
23           Coffee Shop                  Yoga Studio   
24                  Café                  Coffee Shop   
25          Dessert Shop                  Pizza Place   
27           Coffee Shop             Sushi Restaurant   
28              Pharmacy                         Park   
29            Restaurant        Vietnamese Restaurant   
30           Coffee Shop               Breakfast Spot   
31    Mexican Restaurant                         Café   
32             Bookstore  Eastern European Restaurant   
33                  Café                          Pub   
34                 Trail                          Pub   
35    Italian Restaurant                  Coffee Shop   
36           Pizza

## Cluster 2 - Gardens

In [33]:
tor_merged.loc[tor_merged['Cluster Labels'] == 1, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
20  Central Toronto               1                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
20         Women's Store      Department Store           Event Space   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
20  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
20   Dumpling Restaurant            Donut Shop       Doner Restaurant   

   11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
20                Dog Run    Distribution Center         Discount Store   

   14th Most Common Venue 15th Most Common Venue  
20                  Diner           Dessert Shop

## Cluster 3 - Shopping Area with Playgrounds

In [34]:
tor_merged.loc[tor_merged['Cluster Labels'] == 2, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
26  Central Toronto               2            Playground   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
26         Women's Store      Department Store           Event Space   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
26  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
26   Dumpling Restaurant            Donut Shop       Doner Restaurant   

   11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
26                Dog Run    Distribution Center         Discount Store   

   14th Most Common Venue 15th Most Common Venue  
26                  Diner           Dessert Shop

## Cluster 4 - Parks

In [35]:
tor_merged.loc[tor_merged['Cluster Labels'] == 3, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14  Downtown Toronto               3                  Park   
22   Central Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14            Playground                 Trail         Deli / Bodega   
22         Jewelry Store          Home Service                 Trail   

   5th Most Common Venue        6th Most Common Venue  \
14  Ethiopian Restaurant            Electronics Store   
22      Sushi Restaurant  Eastern European Restaurant   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
14  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   
22          Dumpling Restaurant     Electronics Store      Department Store   

   10th Most Common Venue 11th Most Common Venue 12th Most Common Venue  \
14       Doner Restaurant                Dog Run    Distribution Center   
22       Doner Restaurant                Dog Run    Distribution Center   

   13th Most Common Venue 14th Most Common Venue 15th Most Common Venue  
14         Discount Store                  Diner           Dessert Shop  
22   Ethiopian Restaurant         Discount Store                  Diner

## Cluster 5 - Parks near Bus Links

In [36]:
tor_merged.loc[tor_merged['Cluster Labels'] == 4, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19              Bus Line           Swim School    Falafel Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
19           Event Space  Ethiopian Restaurant     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
19  Eastern European Restaurant   Dumpling Restaurant             Donut Shop   

   11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
19       Doner Restaurant                Dog Run    Distribution Center   

   14th Most Common Venue 15th Most Common Venue  
19         Discount Store                  Diner